src/realm/sync/noinst/client_history_impl.cpp

```diff
@@ -359,128 +359,161 @@ void ClientHistory::integrate_server_changesets(const SyncProgress& progress,
+                                   util::format("Failed to parse received changeset: %1", e.what()));
     }
 
-    // downloaded_bytes always contains the total number of downloaded bytes
-    // from the Realm. downloaded_bytes must be persisted in the Realm, since
-    // the downloaded changesets are trimmed after use, and since it would be
-    // expensive to traverse the entire history.
-    Array& root = m_arrays->root;
-    auto downloaded_bytes =
-        std::uint_fast64_t(root.get_as_ref_or_tagged(s_progress_downloaded_bytes_iip).get_as_int());
-    downloaded_bytes += downloaded_bytes_in_message;
-    root.set(s_progress_downloaded_bytes_iip, RefOrTagged::make_tagged(downloaded_bytes)); // Throws
-
-    // The reason we can use the `origin_timestamp`, and the `origin_file_ident`
-    // from the last incoming changeset, and ignore all the other changesets, is
-    // that these values are actually irrelevant for changesets of remote origin
-    // stored in the client-side history (for now), except that
-    // `origin_file_ident` is required to be nonzero, to mark it as having been
-    // received from the server.
-    const Changeset& last_changeset = changesets.back();
-    HistoryEntry entry;
-    entry.origin_timestamp = last_changeset.origin_timestamp;
-    entry.origin_file_ident = last_changeset.origin_file_ident;
-    entry.remote_version = last_changeset.version;
-    entry.changeset = BinaryData(assembled_transformed_changeset.data(), assembled_transformed_changeset.size());
-
-    // m_changeset_from_server is picked up by prepare_changeset(), which then
-    // calls add_sync_history_entry(). prepare_changeset() is called as a result
-    // of committing the current transaction even in the "short-circuited" mode,
-    // because replication isn't disabled.
-    m_changeset_from_server_owner = std::move(assembled_transformed_changeset);
-    REALM_ASSERT(!m_changeset_from_server);
-    m_changeset_from_server = entry;
-
-    // During the bootstrap phase in flexible sync, the server sends multiple download messages with the same
-    // synthetic server version that represents synthetic changesets generated from state on the server.
-    if (batch_state == DownloadBatchState::LastInBatch) {
-        update_sync_progress(progress, downloadable_bytes); // Throws
-    }
+    version_type new_version = 0;
+    constexpr size_t max_number_of_transactions = 10;
+    // Maximum number of subsets we split the `incoming_changesets` into.
+    // We aim that each subset of changesets has the same number of instructions,
+    // but this may not always be possible.
+    int subsets_to_integrate = split_changesets ? std::min(max_number_of_transactions, num_changesets) : 1;
+    std::uint_fast64_t instructions_per_subset = static_cast<std::uint_fast64_t>(
+        std::ceil(static_cast<double>(incoming_instructions_count) / subsets_to_integrate));
```

Comment:
I guess that's possible too. Otherwise, it can happen that when I split the changesets into subsets, there will be some leftover changesets which should be part of the last subset. To do this I need some additional checks which make the code ugly imo.

Consider this case: If we have 6 changesets with the following instruction count distribution: [3, 3, 1, 1, 1, 1], total_intructions = 10. If we were to split it into 3 subsets, then subset_instructions = 10 / 3 = 3. The subsets, as per current code, would be: [3], [3], [1,1,1], leaving the last changeset outside of any subset.